frontend/components/search/InstantSearch.tsx

```diff
@@ -0,0 +1,130 @@
+import Autocomplete from "@material-ui/core/Autocomplete";
+import TextField from "@material-ui/core/TextField";
+import { makeStyles } from "@material-ui/styles";
+import React, { FC, useContext, useRef, useEffect, useState, useCallback } from "react";
+import { SearchFormContext } from "./SearchForm";
+import axios, { AxiosRequestConfig, CancelToken } from "axios";
+import { throttle } from "throttle-debounce";
+
+const useStyles = makeStyles({
+  root: {
+    "& .MuiChip-root": {
+      margin: 1,
+      fontSize: "11px",
+      height: "20px",
+      "& .MuiChip-label": {
+        paddingLeft: "6px",
+        paddingRight: "8px",
+      },
+      "& .MuiChip-deleteIcon": {
+        height: "18px",
+        width: "18px",
+        marginRight: "2px",
+      },
+    },
+  },
+});
+
+type Option = Record<string, string | number>;
+
+interface InstantSearchProps {
+  label: string;
+  name: string;
+  getDataForInput: (input: string, config: AxiosRequestConfig) => Promise<Option[]>;
+  getInitialValue: (ids: number[]) => Promise<Option[]>;
+  labelKey?: string;
+  idKey?: string;
+  minimumSearchLength?: number;
+  throttleDelay?: number;
+}
+
+const InstantSearch: FC<InstantSearchProps> = ({
+  label,
+  name,
+  getDataForInput,
+  getInitialValue,
+  labelKey,
+  idKey = "id",
+  minimumSearchLength = 1,
+  throttleDelay = 250,
+}: InstantSearchProps) => {
+  const classes = useStyles();
+  const { state, setState, disabled } = useContext(SearchFormContext);
+  const [options, setOptions] = useState([]);
+  const [selectedOptions, setSelectedOptions] = useState([]);
+  const [inputValue, setInputValue] = useState("");
+
+  // load labels for values initially loaded from url parameters
+  useEffect(() => {
+    let initialValue: number | number[] = state[name];
+    if (!Array.isArray(initialValue)) {
+      initialValue = [initialValue];
+    }
+    if (initialValue.length === 0) return;
+
+    getInitialValue(initialValue).then((options) => {
+      setSelectedOptions((prevState) => [...prevState, ...options]);
+    });
+  }, []); // eslint-disable-line
+
+  const handleValueChange = (event, values) => {
+    setState((prevState) => ({ ...prevState, [name]: values.map((value) => value[idKey]) }));
+    setSelectedOptions(values);
+  };
+
+  // Throttling behavior only works if the same instance of
+  // the throttled callback is used, so we use `useCallback`.
+  // eslint-disable-next-line
+  const getThrottledDataForInput = useCallback(
+    throttle(throttleDelay, (...args: Parameters<typeof getDataForInput>) =>
```

Comment:
Cool!  What is the throttling behavior?

Reply:
It forces a delay between API requests for efficiency reasons. Not super important right now, but the structure is in place for when application usage increases.